core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -365,7 +367,16 @@ class GroupCoordinator(val brokerId: Int,
 
       case Some(group) =>
         group.inLock {
-          if (group.is(Dead) || !group.has(memberId)) {
+          // if a pending member is leaving, it needs to be removed from the pending list, heartbeat cancelled
+          // and if necessary, prompt a JoinGroup completion.
+          if (group.isPendingMember(memberId)) {
+            group.removePendingMember(memberId)
+            heartbeatPurgatory.cancelForKey(MemberKey(groupId, memberId))
+            if (group.is(PreparingRebalance)) {
```

Comment:
We have two separate cases in which we now remove the pending member and check for rebalance completion. I think it might be worth having a `removePendingMemberAndUpdateGroup` for this logic to go along with `removeMemberAndUpdateGroup`.